File: antd-kotlin/lib/src/main/kotlin/com/autonomi/sdk/Exceptions.kt
package com.autonomi.sdk

import io.grpc.StatusException
import io.grpc.StatusRuntimeException
import io.grpc.Status
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

open class AntdException(message: String, val statusCode: Int = 0) : Exception(message)

class NotFoundException(message: String, statusCode: Int = 404) : AntdException(message, statusCode)
class AlreadyExistsException(message: String, statusCode: Int = 409) : AntdException(message, statusCode)
class ForkException(message: String, statusCode: Int = 409) : AntdException(message, statusCode)
class BadRequestException(message: String, statusCode: Int = 400) : AntdException(message, statusCode)
class PaymentException(message: String, statusCode: Int = 402) : AntdException(message, statusCode)
open class NetworkException(message: String, statusCode: Int = 502) : AntdException(message, statusCode)
class TooLargeException(message: String, statusCode: Int = 413) : AntdException(message, statusCode)
class InternalException(message: String, statusCode: Int = 500) : AntdException(message, statusCode)
class ServiceUnavailableException(message: String, statusCode: Int = 503) : AntdException(message, statusCode)

/**
 * A finalize stored some chunks while others remained unstored after the
 * daemon's retries (HTTP 502 with `code: "PARTIAL_UPLOAD"`; gRPC ABORTED).
 * The on-chain payment persists and the stored chunks stay on the network.
 * How to finish the upload depends on [retryable] and [retentionKnown]
 * ([retryable] implies [retentionKnown]):
 *
 * - [retryable]: the daemon kept the paid attempt (payment proofs +
 *   unstored chunks) under the same `upload_id`. Call the **same** finalize
 *   method again with the same `upload_id` and payment artefacts to store
 *   the remainder against the same payment — no re-prepare, no second
 *   signature, no double payment. Bound the loop: a persistent failure
 *   throws this exception on every call, so cap the attempts and treat a
 *   [chunksFailed] that stops shrinking as stuck. The retained attempt
 *   expires with the daemon's pending-upload TTL.
 * - [retentionKnown] and not [retryable]: the daemon confirmed it kept
 *   nothing (e.g. a merkle finalize with deliberately unpaid batches).
 *   Re-prepare the same content; already-stored chunks are skipped, so the
 *   retry pays only for the missing remainder.
 * - not [retentionKnown]: retention is unknown, and the daemon may still
 *   hold the paid attempt (it records the resume handle before it returns
 *   the error). Stop automatic recovery, keep the `upload_id` and the
 *   original payment artefacts (tx hashes / quote data), and reconcile
 *   before re-preparing or paying again. Never pay again on this signal
 *   alone. Daemons older than 0.14.0 never send `retryable`, so their REST
 *   partial uploads always read as unknown.
 *
 * Extends [NetworkException] because a `PARTIAL_UPLOAD` arrives as a 502,
 * which this SDK has always mapped to [NetworkException]; existing
 * `catch (e: NetworkException)` blocks keep working and can narrow with
 * `is PartialUploadException` when they want the counts and flags. It is
 * not a [ForkException]: over gRPC a partial upload's ABORTED used to map
 * there, and the daemon sends ABORTED only for PARTIAL_UPLOAD, so code that
 * caught [ForkException] around a gRPC finalize should catch
 * [PartialUploadException] instead.
 *
 * Over REST the counts and flags come from the structured error body. Each
 * count is read only from a JSON number holding a non-negative integer;
 * anything else (missing, a quoted number, a negative, fractional or
 * out-of-range number, an array or an object) reads as zero.
 * [retentionKnown] is true when the body's `retryable` is a JSON boolean
 * (`true` or `false`), and [retryable] when it is the literal `true`;
 * missing, `null`, a quoted `"true"`, a number, an array or an object reads
 * as unknown and not retryable. The mapper never throws.
 *
 * Over gRPC they are parsed from the status description
 * (`Partial upload: S/T chunks stored, F failed ...`, with a "paid attempt
 * retained" hint when retryable). [retentionKnown] is true only when the
 * description starts with that layout and all three counts convert to a
 * [Long]; the hint then decides [retryable]. On a layout mismatch or a count
 * that does not convert, all three counts are zero and both flags are false
 * even if the hint is present, so a caller never loops on, or pays again
 * for, a message the SDK could not read.
 *
 * See `docs/external-signer-flow.md` §6 ("Retry a partial store") for the
 * daemon-side contract.
 */
class PartialUploadException(
    message: String,
    val chunksStored: Long = 0,
    val chunksFailed: Long = 0,
    val totalChunks: Long = 0,
    val retryable: Boolean = false,
    retentionKnown: Boolean = retryable,
    statusCode: Int = 502,
) : NetworkException(message, statusCode) {
    /**
     * Whether the daemon's answer on retaining the paid attempt was read:
     * `true` when it confirmed either way, `false` when retention is
     * unknown. Always `true` when [retryable] is.
     */
    val retentionKnown: Boolean = retentionKnown || retryable
}

internal object ExceptionMapping {

    /** Machine-readable `code` the daemon sets on a partial-store finalize. */
    private const val PARTIAL_UPLOAD_CODE = "PARTIAL_UPLOAD"

    /**
     * Fixed text every PARTIAL_UPLOAD message opens with (`antd/src/error.rs`).
     * Over gRPC it is the only thing that distinguishes a partial upload from
     * any other ABORTED status, so [fromGrpcStatus] gates on the description
     * starting with it (anchored, matching the antd-rust client).
     */
    private const val PARTIAL_UPLOAD_PREFIX = "Partial upload:"

    /**
     * Fixed prefix of the daemon's PARTIAL_UPLOAD message, anchored at the
     * start like [isPartialUploadMessage]:
     * `Partial upload: <stored>/<total> chunks stored, <failed> failed`.
     */
    private val partialUploadCounts = Regex("^$PARTIAL_UPLOAD_PREFIX (\\d+)/(\\d+) chunks stored, (\\d+) failed")

    /**
     * Message tail the daemon appends when it kept the paid attempt for a
     * same-upload_id retry.
     */
    private const val PARTIAL_UPLOAD_RETAINED_HINT = "paid attempt retained"

    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Maps a REST error response onto a typed exception, preferring the
     * machine-readable `code` over the bare HTTP status where they diverge:
     * `PARTIAL_UPLOAD` arrives as a 502 that would otherwise read as a plain
     * [NetworkException]. Every other status keeps its status-based mapping.
     */
    fun fromHttpStatus(statusCode: Int, body: String): AntdException =
        partialUploadFromBody(statusCode, body) ?: fromHttpStatusOnly(statusCode, body)

    /**
     * Returns a [PartialUploadException] when [body] is a JSON error object
     * whose `code` is the string `"PARTIAL_UPLOAD"`, carrying its counts.
     * A boolean `retryable` makes retention known and sets `retryable`;
     * daemons < 0.14.0 never send it, so their retention reads as unknown.
     * Returns null for every other body (non-JSON, non-object, or a `code`
     * that is missing, another value, or not a string), so the caller's
     * status-based mapping applies. Counts go through [count] and
     * `retryable` through [flag]: a value of any other kind (a quoted number
     * or quoted `"true"`, a negative or fractional number, an object or an
     * array) reads as absent, zero or `false`, rather than being coerced or
     * escaping as a serialization error. The daemon's error body is input
     * from the network and must never turn a typed [AntdException] into an
     * [IllegalArgumentException].
     */
    private fun partialUploadFromBody(statusCode: Int, body: String): PartialUploadException? {
        val obj = try {
            json.parseToJsonElement(body) as? JsonObject
        } catch (_: Exception) {
            null
        } ?: return null
        val code = obj.primitive("code")
        if (code == null || !code.isString || code.content != PARTIAL_UPLOAD_CODE) return null
        val retained = obj.flag("retryable")
        return PartialUploadException(
            message = obj.primitive("error")?.takeIf { it.isString }?.content ?: body,
            chunksStored = obj.count("chunks_stored"),
            chunksFailed = obj.count("chunks_failed"),
            totalChunks = obj.count("total_chunks"),
            retryable = retained == true,
            retentionKnown = retained != null,
            statusCode = statusCode,
        )
    }

    /**
     * The value at [key] when it is a JSON primitive (string, number, boolean
     * or null); `null` when the key is absent or holds an object or array.
     * Unlike `JsonElement.jsonPrimitive`, never throws.
     */
    private fun JsonObject.primitive(key: String): JsonPrimitive? = this[key] as? JsonPrimitive

    /**
     * The chunk count at [key]: a JSON number holding a non-negative integer
     * that fits a [Long]. Anything else reads as 0: absent, a JSON string
     * (even `"12"`), a negative, fractional or out-of-range number, `null`,
     * a boolean, an object or an array.
     */
    private fun JsonObject.count(key: String): Long =
        primitive(key)?.takeUnless { it.isString }?.content?.toLongOrNull()?.takeIf { it >= 0 } ?: 0

    /**
     * The boolean at [key]: `true` / `false` only for the JSON literals
     * `true` / `false`, and `null` for anything else: absent, `null`, the
     * string `"true"`, a number, an object or an array.
     */
    private fun JsonObject.flag(key: String): Boolean? =
        when (primitive(key)?.takeUnless { it.isString }?.content) {
            "true" -> true
            "false" -> false
            else -> null
        }

    /**
     * True when [message] starts with the daemon's fixed PARTIAL_UPLOAD
     * prefix. Anchored rather than a containment check so an unrelated
     * ABORTED that merely quotes the phrase is not misreported as a partial
     * upload; the daemon never wraps its own message, so the prefix is
     * always at offset zero when it is a partial upload.
     */
    fun isPartialUploadMessage(message: String): Boolean = message.startsWith(PARTIAL_UPLOAD_PREFIX)

    /**
     * Recovers the chunk counts and the retryable hint from a PARTIAL_UPLOAD
     * message. Used for gRPC, where the status carries no structured detail;
     * REST callers get the body fields instead. `retentionKnown` is true only
     * when the message starts with the count layout and all three counts
     * convert to a [Long]; `retryable` additionally needs the "paid attempt
     * retained" hint. The layout accepts any run of digits, so a count past
     * [Long.MAX_VALUE] matches but does not convert. On a layout mismatch or
     * a failed conversion all three counts are zero and both flags are false,
     * even when the hint is present: a retry loop that cannot see
     * [PartialUploadException.chunksFailed] shrinking has no way to tell
     * progress from a stuck upload, and an unread message says nothing about
     * what the daemon kept. Whether the message is a partial upload at all is
     * decided by [isPartialUploadMessage].
     */
    fun partialUploadFromMessage(message: String): PartialUploadException {
        val counts = partialUploadCounts.find(message)?.groupValues?.let { g ->
            val stored = g[1].toLongOrNull()
            val total = g[2].toLongOrNull()
            val failed = g[3].toLongOrNull()
            if (stored != null && total != null && failed != null) Triple(stored, total, failed) else null
        }
        return PartialUploadException(
            message = message,
            chunksStored = counts?.first ?: 0,
            totalChunks = counts?.second ?: 0,
            chunksFailed = counts?.third ?: 0,
            retryable = counts != null && message.contains(PARTIAL_UPLOAD_RETAINED_HINT),
            retentionKnown = counts != null,
        )
    }

    private fun fromHttpStatusOnly(statusCode: Int, body: String): AntdException = when (statusCode) {
        400 -> BadRequestException(body, statusCode)
        402 -> PaymentException(body, statusCode)
        404 -> NotFoundException(body, statusCode)
        409 -> AlreadyExistsException(body, statusCode)
        413 -> TooLargeException(body, statusCode)
        500 -> InternalException(body, statusCode)
        502 -> NetworkException(body, statusCode)
        503 -> ServiceUnavailableException(body, statusCode)
        else -> AntdException(body, statusCode)
    }

    fun fromGrpcStatus(ex: StatusRuntimeException): AntdException {
        val detail = ex.status.description ?: ex.message ?: "Unknown error"
        return when (ex.status.code) {
            Status.Code.NOT_FOUND -> NotFoundException(detail)
            Status.Code.ALREADY_EXISTS -> AlreadyExistsException(detail)
            // PARTIAL_UPLOAD: some chunks stored, some still unstored after
            // retries. The counts and the "paid attempt retained" hint ride
            // the status description over gRPC (no structured detail yet),
            // so parse them best-effort to match the REST client's typed
            // exception. The daemon's message always opens with the fixed
            // "Partial upload:" prefix, so gate on the description starting
            // with it; any other ABORTED keeps the pre-existing
            // conflicting-update mapping instead of being misreported as a
            // partial upload.
            Status.Code.ABORTED ->
                if (isPartialUploadMessage(detail)) partialUploadFromMessage(detail) else ForkException(detail)
            Status.Code.INVALID_ARGUMENT -> BadRequestException(detail)
            Status.Code.FAILED_PRECONDITION -> PaymentException(detail)
            Status.Code.UNAVAILABLE -> NetworkException(detail)
            Status.Code.RESOURCE_EXHAUSTED -> TooLargeException(detail)
            Status.Code.INTERNAL -> InternalException(detail)
            else -> AntdException(detail, ex.status.code.value())
        }
    }
}

File: antd-csharp/Antd.Sdk/Exceptions.cs
using System.Globalization;
using System.Net;
using System.Text.Json;
using System.Text.RegularExpressions;
using Grpc.Core;

namespace Antd.Sdk;

public class AntdException : Exception
{
    public int StatusCode { get; }

    public AntdException(string message, int statusCode = 0)
        : base(message) => StatusCode = statusCode;
}

public class NotFoundException : AntdException
{
    public NotFoundException(string message, int statusCode = 404)
        : base(message, statusCode) { }
}

public class AlreadyExistsException : AntdException
{
    public AlreadyExistsException(string message, int statusCode = 409)
        : base(message, statusCode) { }
}

public class ForkException : AntdException
{
    public ForkException(string message, int statusCode = 409)
        : base(message, statusCode) { }
}

public class BadRequestException : AntdException
{
    public BadRequestException(string message, int statusCode = 400)
        : base(message, statusCode) { }
}

public class PaymentException : AntdException
{
    public PaymentException(string message, int statusCode = 402)
        : base(message, statusCode) { }
}

public class NetworkException : AntdException
{
    public NetworkException(string message, int statusCode = 502)
        : base(message, statusCode) { }
}

/// <summary>
/// A finalize stored some chunks while others remained unstored after the
/// daemon's retries (HTTP 502 with <c>code: "PARTIAL_UPLOAD"</c>; gRPC
/// <c>ABORTED</c> whose status detail starts with the daemon's fixed
/// <c>Partial upload:</c> prefix). The on-chain payment persists and the
/// stored chunks stay on the network. How to finish the upload depends on
/// <see cref="Retryable"/>:
/// <list type="bullet">
/// <item><description>
/// <see cref="Retryable"/> is <c>true</c>: the daemon kept the paid attempt
/// (payment proofs plus the unstored chunks) under the same <c>upload_id</c>.
/// Call the same finalize method again with the same arguments to store the
/// remainder against the same payment: no re-prepare, no second signature,
/// no double payment. Bound that loop: a persistent failure throws this
/// exception on every call, so cap the attempts and treat a
/// <see cref="ChunksFailed"/> that stops shrinking as stuck. The retained
/// attempt expires with the daemon's pending-upload TTL. (antd 0.14.0 and
/// later; older daemons never send the flag, so it reads <c>false</c> and the
/// re-prepare path applies.)
/// </description></item>
/// <item><description>
/// <see cref="Retryable"/> is <c>false</c>: nothing was retained (a merkle
/// finalize with deliberately unpaid batches, or an older daemon).
/// Re-preparing the same content skips already-stored chunks, so a retry
/// pays only for the missing remainder.
/// </description></item>
/// </list>
/// Extends <see cref="NetworkException"/> because a partial upload has
/// always arrived as a 502, so existing <c>catch (NetworkException)</c>
/// blocks keep matching; catch this type first to branch on the counts.
/// Over REST the counts and flag come from the structured error body; a
/// field that is missing or not of the expected JSON kind reads as zero or
/// <c>false</c>. Over gRPC they are parsed best-effort from the status
/// detail (<c>Partial upload: S/T chunks stored, F failed ...</c>, with a
/// <c>paid attempt retained</c> hint when retryable). <see cref="Retryable"/>
/// is <c>true</c> only when all three counts parse and the hint is present:
/// a prefixed detail whose counts do not match or do not convert (an
/// overflow, non-ASCII digits) leaves all three counts zero and
/// <see cref="Retryable"/> <c>false</c>, even with the hint, because a retry
/// loop that cannot watch <see cref="ChunksFailed"/> shrink cannot tell
/// progress from a stuck upload.
/// An <c>ABORTED</c> whose detail does not start with the prefix (including
/// one that quotes it further in) is a <see cref="ForkException"/>, as
/// before.
/// See docs/external-signer-flow.md, section 6.
/// </summary>
public class PartialUploadException : NetworkException
{
    /// <summary>Chunks the daemon confirmed stored before it gave up.</summary>
    public ulong ChunksStored { get; }

    /// <summary>Chunks still unstored after the daemon's own retries.</summary>
    public ulong ChunksFailed { get; }

    /// <summary>Chunks the finalize was asked to store in total.</summary>
    public ulong TotalChunks { get; }

    /// <summary>
    /// <c>true</c> when the daemon retained the paid attempt under the same
    /// <c>upload_id</c>, so the same finalize call can be repeated to store
    /// the remainder against the same payment. Defaults to <c>false</c> when
    /// the daemon did not send the flag (antd before 0.14.0). Over gRPC it is
    /// also <c>false</c> whenever the counts in the status detail could not
    /// be parsed.
    /// </summary>
    public bool Retryable { get; }

    public PartialUploadException(
        string message,
        ulong chunksStored,
        ulong chunksFailed,
        ulong totalChunks,
        bool retryable,
        int statusCode = 502)
        : base(message, statusCode)
    {
        ChunksStored = chunksStored;
        ChunksFailed = chunksFailed;
        TotalChunks = totalChunks;
        Retryable = retryable;
    }
}

public class ServiceUnavailableException : AntdException
{
    public ServiceUnavailableException(string message, int statusCode = 503)
        : base(message, statusCode) { }
}

public class TooLargeException : AntdException
{
    public TooLargeException(string message, int statusCode = 413)
        : base(message, statusCode) { }
}

public class InternalException : AntdException
{
    public InternalException(string message, int statusCode = 500)
        : base(message, statusCode) { }
}

internal static partial class ExceptionMapping
{
    /// <summary>Machine-readable <c>code</c> the daemon sets on a partial upload.</summary>
    private const string PartialUploadCode = "PARTIAL_UPLOAD";

    /// <summary>
    /// Message tail the daemon appends when it kept the paid attempt for a
    /// same-<c>upload_id</c> retry.
    /// </summary>
    private const string PartialUploadRetainedHint = "paid attempt retained";

    /// <summary>
    /// Fixed text every PARTIAL_UPLOAD message from the daemon opens with.
    /// Over gRPC it is the only marker that distinguishes a partial upload
    /// from any other <c>ABORTED</c> status, so the mapping gates on the
    /// status detail starting with it (anchored, as in antd-rust).
    /// </summary>
    private const string PartialUploadPrefix = "Partial upload:";

    /// <summary>
    /// Count layout of the daemon's PARTIAL_UPLOAD message, anchored at its
    /// start: <c>Partial upload: &lt;stored&gt;/&lt;total&gt; chunks stored, &lt;failed&gt; failed</c>.
    /// ASCII digits only: .NET's <c>\d</c> also matches other Unicode decimal
    /// digits, which the daemon never sends.
    /// </summary>
    [GeneratedRegex(@"\APartial upload: ([0-9]+)/([0-9]+) chunks stored, ([0-9]+) failed")]
    private static partial Regex PartialUploadCounts();

    public static AntdException FromHttpStatus(HttpStatusCode status, string body)
    {
        var code = (int)status;
        // Prefer the machine-readable `code` over the bare status where they
        // diverge: PARTIAL_UPLOAD arrives as a 502 that would otherwise read
        // as a generic NetworkException. Every other code keeps the
        // status-based mapping below.
        if (TryParsePartialUploadBody(body, code, out var partial))
            return partial!;
        return code switch
        {
            400 => new BadRequestException(body, code),
            402 => new PaymentException(body, code),
            404 => new NotFoundException(body, code),
            409 => new ForkException(body, code),
            413 => new TooLargeException(body, code),
            500 => new InternalException(body, code),
            502 => new NetworkException(body, code),
            503 => new ServiceUnavailableException(body, code),
            _ => new AntdException(body, code),
        };
    }

    /// <summary>
    /// Recognises the daemon's structured PARTIAL_UPLOAD body
    /// (<c>error</c>, <c>code</c>, <c>chunks_stored</c>, <c>chunks_failed</c>,
    /// <c>total_chunks</c>, <c>retryable</c>). <c>retryable</c> is absent on
    /// daemons before 0.14.0 and then reads <c>false</c>. A body that is not
    /// a readable JSON object, or whose <c>code</c> is not the string
    /// <c>"PARTIAL_UPLOAD"</c>, is left to the status-based mapping. A count
    /// or flag of the wrong JSON kind reads as zero / <c>false</c>, and an
    /// <c>error</c> that is not a decodable string falls back to the raw
    /// body. The body is network input, so this never throws.
    /// </summary>
    private static bool TryParsePartialUploadBody(string body, int statusCode, out PartialUploadException? partial)
    {
        partial = null;
        var trimmed = body?.TrimStart() ?? "";
        if (trimmed.Length == 0 || trimmed[0] != '{')
            return false;
        try
        {
            using var doc = JsonDocument.Parse(trimmed);
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return false;
            if (ReadString(root, "code") != PartialUploadCode)
                return false;

            var message = ReadString(root, "error") ?? body!;
            var retryable = root.TryGetProperty("retryable", out var r) && r.ValueKind == JsonValueKind.True;
            partial = new PartialUploadException(
                message,
                ReadCount(root, "chunks_stored"),
                ReadCount(root, "chunks_failed"),
                ReadCount(root, "total_chunks"),
                retryable,
                statusCode);
            return true;
        }
        catch (Exception e) when (e is JsonException or InvalidOperationException or ArgumentException)
        {
            // JsonException: malformed JSON. InvalidOperationException: a
            // property name whose escape is not valid UTF-16 (a lone
            // surrogate), which System.Text.Json throws on when it has to
            // unescape the name to compare it. ArgumentException: input text
            // that is not valid UTF-16. None is a body this mapper can read,
            // so the status-based mapping applies.
            return false;
        }
    }

    /// <summary>
    /// The value at <paramref name="name"/> when it is a JSON string that
    /// decodes; <c>null</c> when it is absent, another JSON kind, or an escape
    /// that is not valid UTF-16 (a lone surrogate), on which
    /// <see cref="JsonElement.GetString"/> would throw.
    /// </summary>
    private static string? ReadString(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var el) || el.ValueKind != JsonValueKind.String)
            return null;
        try
        {
            return el.GetString();
        }
        catch (InvalidOperationException)
        {
            return null;
        }
    }

    private static ulong ReadCount(JsonElement root, string name) =>
        root.TryGetProperty(name, out var el) && el.ValueKind == JsonValueKind.Number && el.TryGetUInt64(out var v)
            ? v
            : 0UL;

    /// <summary>
    /// Recovers the chunk counts and the retryable hint from a PARTIAL_UPLOAD
    /// message. Used for gRPC, where the status carries no structured detail;
    /// REST callers get the body fields instead. <c>Retryable</c> is
    /// <c>true</c> only when the message matches the count layout, all three
    /// counts convert to <see cref="ulong"/>, and the retained hint is
    /// present. On a layout miss or any failed conversion (an overflow) all
    /// three counts are zero and <c>Retryable</c> is <c>false</c> whatever
    /// the hint says: a retry loop that cannot watch the failed count shrink
    /// cannot tell progress from a stuck upload. Whether the message is a
    /// partial upload at all is decided by <see cref="IsPartialUploadMessage"/>.
    /// </summary>
    internal static (ulong Stored, ulong Failed, ulong Total, bool Retryable) ParsePartialUploadMessage(string? message)
    {
        var msg = message ?? "";
        var m = PartialUploadCounts().Match(msg);
        if (!m.Success
            || !TryParseCount(m.Groups[1].Value, out var stored)
            || !TryParseCount(m.Groups[2].Value, out var total)
            || !TryParseCount(m.Groups[3].Value, out var failed))
        {
            return (0, 0, 0, false);
        }
        var retryable = msg.Contains(PartialUploadRetainedHint, StringComparison.Ordinal);
        return (stored, failed, total, retryable);
    }

    /// <summary>
    /// Converts one captured run of ASCII digits; <c>false</c> when it does
    /// not fit in a <see cref="ulong"/>.
    /// </summary>
    private static bool TryParseCount(string digits, out ulong value) =>
        ulong.TryParse(digits, NumberStyles.None, CultureInfo.InvariantCulture, out value);

    public static AntdException FromGrpcStatus(RpcException ex)
    {
        // The raw status detail, not RpcException.Message: Grpc.Net formats
        // the latter as Status(StatusCode="...", Detail="..."), which would
        // defeat the anchored PARTIAL_UPLOAD match below.
        var detail = ex.Status.Detail;
        return ex.StatusCode switch
        {
            Grpc.Core.StatusCode.NotFound => new NotFoundException(detail),
            Grpc.Core.StatusCode.AlreadyExists => new AlreadyExistsException(detail),
            // ABORTED carries the daemon's PARTIAL_UPLOAD (some chunks stored,
            // some still unstored after retries) when the status detail
            // starts with the fixed "Partial upload:" prefix. The counts and
            // the "paid attempt retained" hint ride the detail text over gRPC
            // (no structured detail yet), so parse them best-effort to match
            // the REST client's typed exception. Any other ABORTED, including
            // one that quotes the prefix further into its detail, keeps the
            // pre-existing version-conflict mapping.
            Grpc.Core.StatusCode.Aborted => AbortedFromGrpc(detail),
            Grpc.Core.StatusCode.InvalidArgument => new BadRequestException(detail),
            Grpc.Core.StatusCode.FailedPrecondition => new PaymentException(detail),
            Grpc.Core.StatusCode.Unavailable => new NetworkException(detail),
            Grpc.Core.StatusCode.ResourceExhausted => new TooLargeException(detail),
            Grpc.Core.StatusCode.Internal => new InternalException(detail),
            _ => new AntdException(detail, (int)ex.StatusCode),
        };
    }

    private static AntdException AbortedFromGrpc(string detail)
    {
        if (!IsPartialUploadMessage(detail))
            return new ForkException(detail);
        var (stored, failed, total, retryable) = ParsePartialUploadMessage(detail);
        return new PartialUploadException(detail, stored, failed, total, retryable);
    }

    /// <summary>
    /// <c>true</c> when the gRPC status detail starts with the daemon's fixed
    /// PARTIAL_UPLOAD prefix. Anchored rather than a containment check, so an
    /// unrelated <c>ABORTED</c> that merely quotes the phrase further into its
    /// detail is not misreported as a partial upload. The daemon
    /// sends its PARTIAL_UPLOAD message as the detail unwrapped, so a genuine
    /// partial upload always has the prefix at offset zero. Pass the raw
    /// <c>Status.Detail</c>, not <c>RpcException.Message</c>.
    /// </summary>
    internal static bool IsPartialUploadMessage(string? message) =>
        message?.StartsWith(PartialUploadPrefix, StringComparison.Ordinal) == true;
}
